### PR TITLE
Add debugging instructions for "Reproducibility between runs"

### PR DESCRIPTION
**Pull request by wwwjn**
As titled. To avoid someone with the same issue as me to understand why we need "seed checkpoint"

### CODE REVIEW

File: docs/debugging.md [hunk 1]
```diff
@@ -59,3 +59,46 @@ This will print a structured configuration to `stdout`, allowing you to verify t
 If you encounter jobs that timeout, you'll need to debug them to identify the root cause. To help with this process, we've enabled Flight Recorder, a tool that continuously collects diagnostic information about your jobs.
 When a job times out, Flight Recorder automatically generates dump files on every rank containing valuable debugging data. You can find these dump files in the `job.dump_folder` directory.
 To learn how to analyze and diagnose issues using these logs, follow our step-by-step tutorial [link](https://pytorch.org/tutorials/prototype/flight_recorder_tutorial.html).
+
+## Reproducibility between Runs
+
+When debugging issues with multi-dimensional parallelism (combinations of FSDP, TP, PP, CP), ensuring reproducible behavior is crucial for isolating and fixing problems. TorchTitan provides several mechanisms to achieve deterministic training runs:
```
Comment (tianyu-l) [inline docs/debugging.md]:
```suggestion
When debugging issues with multi-dimensional parallelism (combinations of FSDP, TP, PP, CP, EP), ensuring reproducible behavior is crucial for isolating and fixing problems. `torchtitan` provides several mechanisms to achieve deterministic training runs:
```

File: docs/debugging.md [hunk 2]
```diff
@@ -59,3 +59,46 @@ This will print a structured configuration to `stdout`, allowing you to verify t
...
+
+### Deterministic Mode
+
+Enable deterministic algorithms to ensure bit-for-bit reproducibility across runs:
```
Comment (tianyu-l) [inline docs/debugging.md:84]: maybe add "under the same training config"?
E.g. if you change parallelism this can't be true

File: docs/debugging.md [hunk 3]
```diff
@@ -59,3 +59,46 @@ This will print a structured configuration to `stdout`, allowing you to verify t
...
+CONFIG_FILE="./train_configs/debug_model.toml" ./run_train.sh --training.deterministic
+```
+
+**What it does:**
+- Forces all CUDA operations to use deterministic algorithms
+- Disables CuDNN benchmarking and enables deterministic mode
+- Sets deterministic workspace configuration for CuBLAS operations
+- **Note:** This will significantly reduce training performance but ensures exact reproducibility
+
+### Seed Configuration
```
Comment (tianyu-l) [inline docs/debugging.md]: let's put this section first, before deterministic, following the order in https://docs.pytorch.org/docs/stable/notes/randomness.html
We could include this link.

File: docs/debugging.md [hunk 4]
```diff
@@ -59,3 +59,46 @@ This will print a structured configuration to `stdout`, allowing you to verify t
...
+
+Set consistent random seeds across all parallelism dimensions:
+
+```bash
+CONFIG_FILE="./train_configs/debug_model.toml" ./run_train.sh --training.seed 42
+```
+
+**Seed behavior with parallelism:**
+- **Data Parallel (DP/FSDP):** All ranks use the same seed for model initialization
+- **Tensor Parallel (TP):** All TP ranks use the same seed for consistent weight sharding
```
Comment (tianyu-l) [inline docs/debugging.md]: yes, all ranks will use the same seed; however, for FSDP and TP, DTensor will do special RNG management to make sure a Replicate tensor get the same init across ranks, but a Shard tensor get "random"-like init across ranks.

File: docs/debugging.md [hunk 5]
```diff
@@ -59,3 +59,46 @@ This will print a structured configuration to `stdout`, allowing you to verify t
...
+- **Pipeline Parallel (PP):** Each PP stage gets a different seed to ensure different dropout patterns
+- **Sequence Parallel:** Maintains consistent seeding across sequence-sharded dimensions
```
Comment (tianyu-l) [inline docs/debugging.md]: what's this? CP?
If so, for parameters, `dp_shard` mesh and `cp` mesh together form the FSDP sharding group. So if dp_shard=2, cp=2, the parameter sharding are the same as dp_shard=4, cp=1.

File: docs/debugging.md [hunk 6]
```diff
@@ -59,3 +59,46 @@ This will print a structured configuration to `stdout`, allowing you to verify t
...
+
+### Seed-Checkpoint-based Reproducibility
+
+For experimental runs with TP/PP/EP, we need use `seed_checkpoint` to ensure we initialize the model with the same weights on each rank. This is because in torchtitan/train.py, the model is sharded first, and then initialized weights on each rank seprately. So it's not equivalent to initialize the model on rank 0 and then shard it. An workaround is to use `seed_checkpoint` to make sure the model on each rank with the same weights, becuase the weights are loaded from the same checkpoint.
```
Comment (tianyu-l) [inline docs/debugging.md]:
```suggestion
For multiple experimental runs with different parallelism configs, we need to use a "seed" checkpoint to ensure model initializations are the same across runs. This is because in `torchtitan/train.py`, the model parameters are sharded first, and then have their weights initialized on each rank separately. As a result, it is not equivalent to initialize the model on one rank and then shard it. Using a seed checkpoint helps different runs load the same model weights from checkpoint -- DCP resharding will make sure the loaded weights are sharded correctly according to the parallelism configs.
```

File: docs/debugging.md [hunk 7]
```diff
@@ -59,3 +59,46 @@ This will print a structured configuration to `stdout`, allowing you to verify t
...
+
+
+```bash
+NGPU=1 CONFIG_FILE="./train_configs/debug_model.toml" ./run_train.sh --checkpoint.enable_checkpoint --checkpoint.create_seed_checkpoint --parallelism.data_parallel_replicate_degree 1 --parallelism.data_parallel_shard_degree 1 --parallelism.tensor_parallel_degree 1 --parallelism.pipeline_parallel_degree 1 --parallelism.context_parallel_degree 1
+```
+
+Note: Using seed checkpoint will only make sure the model on each rank initialized with the same weights, but the later training process might not be exactly the same even with the `deterministic` mode (eg, the precision change during parallesim might cause the different results).
```
Comment (tianyu-l) [inline docs/debugging.md]:
```suggestion
Note: Using a seed checkpoint will only make sure a model has same initial weights when configs change, but the training process may not be the same even after setting the seed and the `deterministic` mode, e.g. due to tensor shape change, data precision change, usage of randomness in model code, etc.
```

File: docs/debugging.md [hunk 8]
```diff
@@ -59,3 +59,63 @@ This will print a structured configuration to `stdout`, allowing you to verify t
...
+
+```bash
+CONFIG_FILE="./train_configs/debug_model.toml" ./run_train.sh --training.seed 42
+```
+
+**Seed behavior with parallelism:**
+- **Data Parallel (DP/FSDP):** All ranks use the same seed for model initialization
+- **Tensor Parallel (TP):** All TP ranks use the same seed for consistent weight sharding
+    - Note: For FSDP and TP, DTensor will do special RNG management to make sure a Replicate tensor get the same init across ranks, but a Shard tensor get "random"-like init across ranks.
+- **Pipeline Parallel (PP):** Each PP stage gets a different seed to ensure different dropout patterns
```
Comment (tianyu-l) [inline docs/debugging.md]:
```suggestion
- **Pipeline Parallel (PP):** Each PP stage gets a different seed to ensure different initialization across layers on different PP ranks.
```

File: docs/debugging.md [hunk 9]
```diff
@@ -59,3 +59,63 @@ This will print a structured configuration to `stdout`, allowing you to verify t
...
+
+### Seed Configuration
+Set consistent random seeds across all parallelism dimensions:
+
+```bash
+CONFIG_FILE="./train_configs/debug_model.toml" ./run_train.sh --training.seed 42
+```
+
+**Seed behavior with parallelism:**
+- **Data Parallel (DP/FSDP):** All ranks use the same seed for model initialization
```
Comment (tianyu-l) [inline docs/debugging.md]:
```suggestion
- **Data Parallel (DP/FSDP):** All DP ranks use the same seed.
```

File: docs/debugging.md [hunk 10]
```diff
@@ -59,3 +59,63 @@ This will print a structured configuration to `stdout`, allowing you to verify t
...
+- **Tensor Parallel (TP):** All TP ranks use the same seed for consistent weight sharding
```
Comment (tianyu-l) [inline docs/debugging.md]:
```suggestion
- **Tensor Parallel (TP):** All TP ranks use the same seed.
```

File: docs/debugging.md [hunk 11]
```diff
@@ -59,3 +59,63 @@ This will print a structured configuration to `stdout`, allowing you to verify t
...
+    - Note: For FSDP and TP, DTensor will do special RNG management to make sure a Replicate tensor get the same init across ranks, but a Shard tensor get "random"-like init across ranks.
+- **Pipeline Parallel (PP):** Each PP stage gets a different seed to ensure different dropout patterns
+- **Context Parallel (CP):** All ranks use the same seed for model initialization
```
Comment (tianyu-l) [inline docs/debugging.md]:
```suggestion
- **Context Parallel (CP):** All ranks use the same seed.
```

File: docs/debugging.md [hunk 12]
```diff
@@ -59,3 +59,63 @@ This will print a structured configuration to `stdout`, allowing you to verify t
...
+When debugging issues with multi-dimensional parallelism (combinations of FSDP, TP, PP, CP, EP), ensuring reproducible behavior is crucial for isolating and fixing problems. `torchtitan` provides several mechanisms to achieve deterministic training runs. For more information on ensuring reproducibility and managing randomness in PyTorch, you can refer to the official PyTorch documentation on randomness: [PyTorch Randomness Documentation](https://docs.pytorch.org/docs/stable/notes/randomness.html).
+
+### Seed Configuration
+Set consistent random seeds across all parallelism dimensions:
+
+```bash
+CONFIG_FILE="./train_configs/debug_model.toml" ./run_train.sh --training.seed 42
+```
+
+**Seed behavior with parallelism:**
```
Comment (tianyu-l) [inline docs/debugging.md:76]: we can combine DP/TP/CP as they use the same seed

File: docs/debugging.md [hunk 13]
```diff
@@ -59,3 +59,63 @@ This will print a structured configuration to `stdout`, allowing you to verify t
...
+- **Data Parallel (DP/FSDP):** All ranks use the same seed for model initialization
+- **Tensor Parallel (TP):** All TP ranks use the same seed for consistent weight sharding
+    - Note: For FSDP and TP, DTensor will do special RNG management to make sure a Replicate tensor get the same init across ranks, but a Shard tensor get "random"-like init across ranks.
+- **Pipeline Parallel (PP):** Each PP stage gets a different seed to ensure different dropout patterns
+- **Context Parallel (CP):** All ranks use the same seed for model initialization
+
+
+### Deterministic Mode
+
+Enable deterministic algorithms to ensure bit-for-bit reproducibility across runs:
```
Comment (tianyu-l) [inline docs/debugging.md:84]: let's add this

File: docs/debugging.md [hunk 14]
```diff
@@ -59,3 +59,63 @@ This will print a structured configuration to `stdout`, allowing you to verify t
...
+A common scenario is when you introduce a new parallelism strategy to the model, you need to ensure that the loss curve remains numerically equivalent to the previous parallelism plan, thereby confirming the accuracy of your implementation. To achieve consistent behavior across multiple runs with varying parallelism configurations, it's crucial to make sure dataloader behaves consistently. We need to fix the DP degree (`dp_replicate * dpshard`) to ensure the dataloader operates consistently.
+
+Here's a typical comparison setup (maintaining an overall DP degree of 4):
+- Run #1: FSDP degree = 4
+- Run #2: DDP degree = 2, FSDP dgree = 2, TP degree = 2
+- Run #3: DDP degree = 2, FSDP degree = 2, CP degree = 2, PP degree = 2
+
+To reproduce loss curves across above runs, you'll need to create a seed checkpoint, and then load the same seed checkpoint for all runs to ensure consistent model initialization on each rank. You will also need to set the `deterministic` mode to ensure consistent training behavior.
+
+We also provided an example of verifying the numerical consistency across parallism plans on Llama3 in https://github.com/pytorch/torchtitan/blob/main/docs/converging.md.
```
Comment (tianyu-l) [inline docs/debugging.md]:
```suggestion
We also provided an example of verifying the numerical consistency across parallism configs on Llama 3 in https://github.com/pytorch/torchtitan/blob/main/docs/converging.md.
```

File: docs/debugging.md [hunk 15]
```diff
@@ -59,3 +59,63 @@ This will print a structured configuration to `stdout`, allowing you to verify t
...
+### Example: Reproducing loss curves with different parallelism configs
+
+A common scenario is when you introduce a new parallelism strategy to the model, you need to ensure that the loss curve remains numerically equivalent to the previous parallelism plan, thereby confirming the accuracy of your implementation. To achieve consistent behavior across multiple runs with varying parallelism configurations, it's crucial to make sure dataloader behaves consistently. We need to fix the DP degree (`dp_replicate * dpshard`) to ensure the dataloader operates consistently.
+
+Here's a typical comparison setup (maintaining an overall DP degree of 4):
+- Run #1: FSDP degree = 4
+- Run #2: DDP degree = 2, FSDP dgree = 2, TP degree = 2
+- Run #3: DDP degree = 2, FSDP degree = 2, CP degree = 2, PP degree = 2
+
+To reproduce loss curves across above runs, you'll need to create a seed checkpoint, and then load the same seed checkpoint for all runs to ensure consistent model initialization on each rank. You will also need to set the `deterministic` mode to ensure consistent training behavior.
```
Comment (tianyu-l) [inline docs/debugging.md]:
```suggestion
To reproduce loss curves across above runs, you'll need to create a seed checkpoint, and then load the same seed checkpoint for all runs to ensure consistent model initialization on each rank. You might also need to set the `deterministic` mode to ensure consistent training behavior.
```

File: docs/debugging.md [hunk 16]
```diff
@@ -59,3 +59,63 @@ This will print a structured configuration to `stdout`, allowing you to verify t
...
+Note: Using a seed checkpoint will only make sure a model has same initial weights when configs change, but the training process may not be the same even after setting the seed and the `deterministic` mode, e.g. due to tensor shape change, data precision change, usage of randomness in model code, etc.
+
+### Example: Reproducing loss curves with different parallelism configs
+
+A common scenario is when you introduce a new parallelism strategy to the model, you need to ensure that the loss curve remains numerically equivalent to the previous parallelism plan, thereby confirming the accuracy of your implementation. To achieve consistent behavior across multiple runs with varying parallelism configurations, it's crucial to make sure dataloader behaves consistently. We need to fix the DP degree (`dp_replicate * dpshard`) to ensure the dataloader operates consistently.
+
+Here's a typical comparison setup (maintaining an overall DP degree of 4):
+- Run #1: FSDP degree = 4
+- Run #2: DDP degree = 2, FSDP dgree = 2, TP degree = 2
+- Run #3: DDP degree = 2, FSDP degree = 2, CP degree = 2, PP degree = 2
```
Comment (tianyu-l) [inline docs/debugging.md]: use dp_shard / dp_replicate instead, for the reasons I mentioned during our discussion

File: docs/debugging.md [hunk 17]
```diff
@@ -59,3 +59,63 @@ This will print a structured configuration to `stdout`, allowing you to verify t
...
+```
+
+Note: Using a seed checkpoint will only make sure a model has same initial weights when configs change, but the training process may not be the same even after setting the seed and the `deterministic` mode, e.g. due to tensor shape change, data precision change, usage of randomness in model code, etc.
+
+### Example: Reproducing loss curves with different parallelism configs
+
+A common scenario is when you introduce a new parallelism strategy to the model, you need to ensure that the loss curve remains numerically equivalent to the previous parallelism plan, thereby confirming the accuracy of your implementation. To achieve consistent behavior across multiple runs with varying parallelism configurations, it's crucial to make sure dataloader behaves consistently. We need to fix the DP degree (`dp_replicate * dpshard`) to ensure the dataloader operates consistently.
```
Comment (tianyu-l) [inline docs/debugging.md]:
```suggestion
A common scenario is when you introduce a new parallelism strategy to the model, you need to ensure that the loss curve remains numerically equivalent to the previous parallelism config, thereby confirming the accuracy of your implementation. To achieve consistent behavior across multiple runs with varying parallelism configurations, it's crucial to make sure dataloader behaves consistently. We need to fix the DP degree (`dp_replicate * dpshard`) to ensure the dataloader operates consistently.
```

**Review by tianyu-l — APPROVED** [review]
please address remaining comments
